util/offline_db/get_resources_from_db.py

```diff
@@ -0,0 +1,238 @@
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+"""
+This is a script to get all resources from the database and save them to
+a json file and download all resources to a directory. The resources in
+the resources.json file that is generated will be updated to point to the
+local resources instead of the database resources. This script is used to
+create the offline database. To use the resources.json with gem5 you need
+to update the gem5 config to point to the resources.json file.
+"""
+
+import argparse
+import json
+from urllib import request, parse
+import pathlib
+from typing import List
+import hashlib
+from tqdm import tqdm
+
+
+def get_token(auth_url: str, api_key: str) -> str:
+    """
+    This function gets the token from the database using the api key
+    :param auth_url: Authentication url to use Atlas data API
+    :param api_key: API key to access the database
+    :return: Token to access the database
+    """
+
+    data = {"key": api_key}
+    data = json.dumps(data).encode("utf-8")
+
+    req = request.Request(
+        auth_url, data=data, headers={"content-type": "application/json"}
+    )
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(f"Failed to get token: {e}")
+
+    token = json.loads(response.read().decode("utf-8"))["access_token"]
+    return token
+
+
+def get_all_resources(
+    url: str, data_source: str, collection: str, database: str, token: str
+) -> List:
+    """
+    This function gets all the JSON objects for resources from the database
+    :param url: Database url to use Atlas data API
+    :param data_source: Data source name for the mongoDB database
+    :param collection: Collection name for the mongoDB database
+    :param database: Database name for the mongoDB database
+    :param token: Token to access the database
+
+    :return: List of JSON objects for resources
+    """
+
+    url = f"{url}/action/find"
+    data = {
+        "dataSource": data_source,
+        "collection": collection,
+        "database": database,
+    }
+    data = json.dumps(data).encode("utf-8")
+    headers = {
+        "Authorization": f"Bearer {token}",
+        "Content-Type": "application/json",
+    }
+
+    req = request.Request(url, data=data, headers=headers)
+
+    try:
+        response = request.urlopen(req)
+    except Exception as e:
+        raise Exception(f"Failed to get resources: {e}")
+
+    return json.loads(response.read().decode("utf-8"))["documents"]
+
+
+def save_resources_to_file(resources: List, output: pathlib.Path):
```

Comment:
In cases where it's possible to know can you state, via typing the type of objects the `List` type will consist of. In this case `List[Dict]`.

Also, in this case I think `pathlib.Path` as the type here is overly specific. I'd do `TextIO`. it can be imported `from typing import TextIO`. `TextIO` is a "any text IO", which is what `open` can handle in this case. `Path` of this type so it's fine.

Finally, i think it's a good habit to add a return type. In this case the return type is `None`.


```suggestion
def save_resources_to_file(resources: List[Dict], output: TextIO) -> None:
```